src/Application/exceptions.php

```diff
@@ -44,9 +44,29 @@ class BadRequestException extends \Exception
 	protected $code = 404;
 
 
-	public function __construct($message = '', $code = 0, \Exception $previous = NULL)
+	/** @var Request */
+	private $request;
+
+
+	public function __construct($message = '', $code = 0, \Exception $previous = NULL, Request $request = NULL)
 	{
 		parent::__construct($message, $code < 200 || $code > 504 ? $this->code : $code, $previous);
+		$this->request = $request;
+	}
+
+
+	/**
+	 * @return Request|null
+	 */
+	public function getRequest()
+	{
+		return $this->request;
+	}
+
+
+	public static function fromRequest(Request $request, $message = '', $code = 0, \Exception $previous = NULL)
```

Comment:
@fprochazka Then what do you think is better? The implementation currently proposed or this one? :-)

```php
/**
 * @param string|\Exception $message
 */
public static function fromRequest(Request $request, $message)
{
	$cond = $message instanceof \Exception;
	return new static($cond ? $message->getMessage(), 0, $cond ? $message : null, $request);
}
```